File: 1D_DP/LC_139.py
# Written by: Jian-Yuan Lin
# Date: 2022.12.06

class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        # try bottom up
        n = len(s)
        dp = [False] * (n+1)
        dp[n] = True
        
        for i in range(n-1, -1, -1):
            # compare every possible word in wordDict
            for w in wordDict:
                if i + len(w) <= n and s[i:i+len(w)] == w:
                    dp[i] = dp[i+len(w)]
                if dp[i]:
                    break
        return dp[0]
        """
        Top down + memoization
        mem = {}
        mem[""] = True
        worddict = set(wordDict)
        def dp(s):
            if s in mem:
                return mem[s]
            if s in worddict:
                mem[s] = True
                return mem[s]
            # make split
            mem[s] = False
            for i in range(1, len(s)):
                first, second = s[:i], s[i:]
                if dp(first) and second in worddict:
                    mem[s] = True
                    break
            return mem[s]
        return dp(s)
        """
        

File: 1D_DP/LC_139.cpp
// Written by: Jian-Yuan Lin
// Date: 2022.12.06

class Solution {
public:
    bool dp(string s, unordered_map<string, bool>& mem, unordered_set<string>& seen){
        if (mem.count(s)){
            return mem[s];
        }
        if (seen.count(s)){
            mem[s] = true;
            return mem[s];
        }
        // split the word 
        mem[s] = false;
        for (int i = 1; i < s.size(); i++){
            string first = s.substr(0, i);
            string second = s.substr(i, s.size() - i);
            if (dp(first, mem, seen) && seen.count(second)){
                mem[s] = true;
                break;
            }
        }
        return mem[s];
    }
    bool wordBreak(string s, vector<string>& wordDict) {
        unordered_map<string, bool> mem;
        mem[""] = true;
        unordered_set<string> seen(wordDict.begin(), wordDict.end());

        return dp(s, mem, seen);
    }
};